File: src/crosschat/state.py
from __future__ import annotations

import asyncio
import json
from datetime import datetime, timezone
from typing import Any, Callable

import aiomqtt
import structlog

from crosschat.models import CrossChatServer, CrossChatUser

log = structlog.get_logger()


class CrossChatState:
	_client: aiomqtt.Client | None  # None if it is us. #TODO: verify there is only one with None

	def __init__(self) -> None:
		self.servers: dict[str, CrossChatServer] = {}
		self._own_id: str = ''
		self._client = None
		self._next_seq: int = 1
		self._own_meta: dict = {}
		self._subscribers: dict[str, list[Callable]] = {}
		self._ooc_subscribers: dict[str, list[Callable]] = {}
		self._prefix: str = 'crosschat/'
		self._tg: asyncio.TaskGroup | None = None

	def set_own_id(self, sid: str) -> None:
		self._own_id = sid
		self._ensure_server(sid)

	def _ensure_server(self, sid: str, ensure=True) -> CrossChatServer:
		if sid not in self.servers:
			self.servers[sid] = CrossChatServer(id=sid, _state=self)
			if not ensure:
				log.warning('Created unknown server', id=sid)
		return self.servers[sid]

	def me(self):
		return self.servers[self._own_id]

	def set_status(self, sid: str, started: int) -> None:
		server = self.me()
		server.started = started
		server.online = started > 0

	def set_meta(self, meta: dict) -> None:
		self._own_meta = meta
		server = self.me()
		server.meta = meta

	def get_meta(self) -> dict:
		return self._own_meta

	def set_client(self, client: aiomqtt.Client, prefix: str) -> None:
		self._client = client
		self._prefix = prefix

	async def publish(self, topic: str, payload: str | dict | list, qos: int = 2, retain: bool = False) -> None:
		if self._client is None:
			return
		full_topic = f'{self._prefix}{topic}'
		if not isinstance(payload, str):
			payload = json.dumps(payload)
		try:
			await self._client.publish(full_topic, payload=payload, qos=qos, retain=retain)
		except aiomqtt.exceptions.MqttCodeError:
			if self._client:
				raise
			raise

	def set_task_group(self, tg: asyncio.TaskGroup) -> None:
		self._tg = tg

	def set_state(self, key: str, value: str) -> None:
		server = self.me()
		server.states[key] = value
		if self._client is not None and self._tg is not None:
			self._tg.create_task(self._publish_state(key, value))
		self._notify(server, key, value)

	def subscribe(self, key: str, callback: Callable) -> None:
		if key not in self._subscribers:
			self._subscribers[key] = []
		self._subscribers[key].append(callback)

	def _notify(self, server: CrossChatServer, key: str, value: str) -> None:
		if key in self._subscribers and self._tg is not None:
			for cb in self._subscribers[key]:
				self._tg.create_task(cb(server, key, value))

	async def _publish_state(self, key: str, value: str) -> None:
		await self.publish(f'state/{self._own_id}/{key}', payload=str(value), retain=True)

	def get_or_create_user(self, server_id: str, user_id: int) -> CrossChatUser:
		server = self._ensure_server(server_id)
		if user_id in server.users:
			return server.users[user_id]
		user = CrossChatUser(
			name='',
			first_seen=datetime.now(timezone.utc),
			server=server,
		)
		server.users[user_id] = user
		return user

	async def add_user(self, name: str, extra: dict | None = None) -> int:
		server = self.me()
		return await server.add_user(name, extra)

	async def del_user(self, user_id: int) -> CrossChatUser | None:
		server = self.me()
		return await server.del_user(user_id)

	def subscribe_ooc(self, ooc_name: str, callback: Callable) -> None:
		if ooc_name not in self._ooc_subscribers:
			self._ooc_subscribers[ooc_name] = []
		self._ooc_subscribers[ooc_name].append(callback)

	async def send_ooc(self, target_sid: str, ooc_name: str, payload: Any) -> None:
		await self.publish(f'm/{self._own_id}/{target_sid}/ooc/{ooc_name}', payload=payload)

	async def _notify_ooc(self, server: CrossChatServer, ooc_name: str, payload: str) -> None:
		if ooc_name in self._ooc_subscribers:
			for cb in self._ooc_subscribers[ooc_name]:
				await cb(server, payload, ooc_name)

	def format_status(self) -> str:
		parts = [f'[Own ID: {self._own_id}]', '']
		for sid in sorted(self.servers):
			server = self.servers[sid]
			badge = 'ONLINE' if server.online else 'OFFLINE'
			marker = ' (self)' if sid == self._own_id else ''
			parts.append(f'  {sid}: {badge}{marker}')
			for uid in sorted(server.users):
				user = server.users[uid]
				ts = user.first_seen.strftime('%Y-%m-%d %H:%M:%S UTC')
				parts.append(f'    └ {uid} ({user.name})  (since {ts})')
		if not self.servers:
			parts.append('  (no servers known)')
		return '\n'.join(parts)

File: csharp/State.cs
using System.Text.Json;
using MQTTnet;
using MQTTnet.Protocol;

namespace CrossChat;

public class CrossChatState
{
    public Dictionary<string, CrossChatServer> Servers { get; } = [];
    public string OwnId { get; private set; } = string.Empty;
    public int NextSeq { get; set; } = 1;
    public IMqttClient? Client { get; private set; }

    private string _prefix = "crosschat/";
    private Dictionary<string, List<Func<CrossChatServer, string, string, Task>>> _subscribers = [];
    private Dictionary<string, List<Func<CrossChatServer, string, string, Task>>> _oocSubscribers = [];
    private Dictionary<string, object> _ownMeta = [];

    public void SetOwnId(string sid)
    {
        OwnId = sid;
        EnsureServer(sid);
    }

    public CrossChatServer EnsureServer(string sid)
    {
        if (!Servers.TryGetValue(sid, out var server))
        {
            server = new CrossChatServer(sid, this);
            Servers[sid] = server;
        }
        return server;
    }

    public CrossChatServer Me() => Servers[OwnId];

    public void SetStatus(string sid, long started)
    {
        if (Servers.TryGetValue(sid, out var server))
        {
            server.Started = started;
            server.Online = started > 0;
        }
    }

    public void SetMeta(Dictionary<string, object> meta)
    {
        _ownMeta = meta;
        if (Servers.TryGetValue(OwnId, out var server))
            server.Meta = meta;
    }

    public Dictionary<string, object> GetMeta() => _ownMeta;

    public void SetClient(IMqttClient client, string prefix)
    {
        Client = client;
        _prefix = prefix;
    }

    public async Task Publish(string topic, string payload, int qos = 2, bool retain = false)
    {
        if (Client == null) return;
        var fullTopic = $"{_prefix}{topic}";
        try
        {
            await Client.PublishAsync(new MqttApplicationMessageBuilder()
                .WithTopic(fullTopic)
                .WithPayload(payload)
                .WithQualityOfServiceLevel((MqttQualityOfServiceLevel)qos)
                .WithRetainFlag(retain)
                .Build());
        }
        catch (Exception ex)
        {
            Console.Error.WriteLine($"Publish error on {fullTopic}: {ex.Message}");
        }
    }

    public void SetState(string key, string value)
    {
        var server = Me();
        server.States[key] = value;
        if (Client != null)
        {
            Task.Run(async () =>
            {
                await Publish($"state/{OwnId}/{key}", value, retain: true);
            });
        }
        Notify(server, key, value);
    }

    public void Subscribe(string key, Func<CrossChatServer, string, string, Task> callback)
    {
        if (!_subscribers.TryGetValue(key, out var list))
        {
            list = [];
            _subscribers[key] = list;
        }
        list.Add(callback);
    }

    internal void Notify(CrossChatServer server, string key, string value)
    {
        if (_subscribers.TryGetValue(key, out var list))
        {
            foreach (var cb in list)
            {
                Task.Run(async () => await cb(server, key, value));
            }
        }
    }

    public CrossChatUser GetOrCreateUser(string serverId, int userId)
    {
        var server = EnsureServer(serverId);
        if (server.Users.TryGetValue(userId, out var user))
            return user;
        user = new CrossChatUser
        {
            Name = string.Empty,
            FirstSeen = DateTime.UtcNow,
            Server = server,
        };
        server.Users[userId] = user;
        return user;
    }

    public async Task<int> AddUser(string name, Dictionary<string, object>? extra = null)
    {
        var server = Me();
        return await server.AddUser(name, extra);
    }

    public async Task<CrossChatUser?> DelUser(int userId)
    {
        var server = Me();
        return await server.DelUser(userId);
    }

    public void SubscribeOoc(string oocName, Func<CrossChatServer, string, string, Task> callback)
    {
        if (!_oocSubscribers.TryGetValue(oocName, out var list))
        {
            list = [];
            _oocSubscribers[oocName] = list;
        }
        list.Add(callback);
    }

    public async Task SendOoc(string targetSid, string oocName, object payload)
    {
        var jsonPayload = JsonSerializer.Serialize(payload);
        await Publish($"m/{OwnId}/{targetSid}/ooc/{oocName}", jsonPayload);
    }

    internal async Task NotifyOoc(CrossChatServer server, string oocName, string payload)
    {
        if (_oocSubscribers.TryGetValue(oocName, out var list))
        {
            foreach (var cb in list)
                await cb(server, payload, oocName);
        }
    }

    public string FormatStatus()
    {
        const string GREY = "\x1b[90m";
        const string WHITE = "\x1b[97m";
        const string GREEN = "\x1b[32m";
        const string RED = "\x1b[31m";
        const string ORANGE = "\x1b[33m";
        const string LIGHTBLUE = "\x1b[36m";
        const string R = "\x1b[0m";

        var parts = new List<string>
        {
            $"{WHITE}[Own ID: {OwnId}]{R}\n"
        };

        foreach (var sid in Servers.Keys.OrderBy(k => k))
        {
            var server = Servers[sid];
            var badge = server.Online ? $"{GREEN}ONLINE{R}" : $"{RED}OFFLINE{R}";
            var marker = sid == OwnId ? $"{GREY} (self){R}" : "";

            var active = server.Users.Count;
            var countStr = $"{GREY} ({active}){R}";

            parts.Add($"  {ORANGE}{sid}{R}: {badge}{marker}{countStr}\n");

            foreach (var uid in server.Users.Keys.OrderBy(k => k))
            {
                var user = server.Users[uid];
                var nameStr = $"{WHITE}{(string.IsNullOrEmpty(user.Name) ? "?" : user.Name)}{R}";
                parts.Add($"  {GREEN}\u2611{R} {GREY}#{uid} {R}{nameStr}\n");

                if (user.Extra.Count > 0)
                {
                    var extraKeys = user.Extra.Keys.OrderBy(k => k).ToList();
                    var extras = string.Join(", ", extraKeys.Select(k => $"{LIGHTBLUE}{k}{R}"));
                    parts.Add($"    {GREY}extra: {R}{extras}\n");
                }
            }
        }

        if (Servers.Count == 0)
            parts.Add($"{GREY}  (no servers known){R}\n");

        return string.Concat(parts);
    }
}
